buf/registry/module/v1beta1/vcs_commit.proto

```diff
@@ -16,124 +16,49 @@ syntax = "proto3";
 //
-// This is used in requests.
-message VCSCommitRef {
-  option (buf.registry.priv.extension.v1beta1.message).request_only = true;
-
-  // The fully-qualified name of a VCSCommit within a BSR instance.
-  //
-  // A Name uniquely identifies a VCSCommit.
-  // This is used for requests when a caller only has the VCSCommit hash and not the ID.
-  message Name {
-    // The name of the owner of the Module that contains the VCSCommit, either a user or organization.
-    string owner = 1 [
-      (buf.validate.field).required = true,
-      (buf.validate.field).string.max_len = 32
-    ];
-    // The name of the Module that contains the VCSCommit.
-    string module = 2 [(buf.validate.field).string = {
-      min_len: 2,
-      max_len: 100
-    }];
-    // The hash of the VCSCommit.
-    string hash = 3 [
-      (buf.validate.field).required = true,
-      (buf.validate.field).string.pattern = "^[0-9a-fA-F]{40}$"
-    ];
-  }
-
-  oneof value {
-    option (buf.validate.oneof).required = true;
-    // The id of the VCSCommit.
-    string id = 1 [(buf.validate.field).string.uuid = true];
-    // The fully-qualified name of the VCSCommit.
-    Name name = 2;
-  }
+// Stored within a Commit, not a resource on its own. Commits <-> VCSCommits are many-to-many,
+// we do not enforce any uniqueness principles here, VCSCommits should just be considered
+// metadata on a given Commit. As such, making them their own entity would result in checks
+// we don't want to enforce. It is still split out into a separate message so that we can
+// enforce required checks on certain fields iff VCSCommit is present.
+//
+// TODO: Below is the explanation, but it's a little unsatisfying. Perhaps we revisit making
+// VCSCommit an entity?
+//
+// Multiple BSR commits may have the same content, but different VCS hashes. This costs some
+// storage, but it allows the BSR to more accurately link back to source control. (For example,
+// git commit A introduces a new service, commit B modifies it, and commit C deletes the
+// service and moves the functionality to a different module. We want the BSR commits created
+// from A and C to link back to the correct git commits and PR discussions.)
+//
+// Multiple BSR commits may have the same VCS hash, but different content. This should only
+// happen when users incorrectly hand-specify the VCS hash — and since commits are immutable and
+// can’t be deleted, we don’t want a manually-created commit to forever block a later,
+// fully-correct commit.
+message VCSCommit {
+  // The type of VCS that the Commit originated from.
+  VCSType type = 1 [
+    (buf.validate.field).required = true,
+    (buf.validate.field).enum.defined_only = true
+  ];
+  // The VCS hash of the Commit.
+  string hash = 2 [
+    (buf.validate.field).required = true,
+    (buf.validate.field).string.pattern = "^[0-9a-fA-F]{40}$"
+  ];
+  // The URL of of the commit within the VCS.
+  string url = 3 [(buf.validate.field).string.uri = true];
```

Comment:
@akshayjshah and I debated whether this should be a per-commit URL or a per-module URL. The current proposal in Notion is for this to be a per-module URL (so the current implementation deviates from the proposal, and if we keep this we should update the proposal). The distinction here is relevant if/when enterprises move code hosts (which does happen and is not uncommon). If we store this on a per-commit basis, and if a customer updates their code host but doesn't keep their old code host active and doesn't setup redirects, then we will end up with a bunch of dead links. Having a module level URL template would make this easy to remedy. 

Having said all that.
* Maybe it is ok to have old commits point to dead links?
* Maybe it is ok to see if enterprises keep old code hosts running or setup redirects in practice?
* Maybe it is ok we punt on this until there is an actual enterprise with an actual problem, and we could always add a module level override in the future (less config now)?

Semantically, all we really care to capture is a link back to the code host. Maybe this is better named CodeHostLink instead of VCS commit. Another argument for this is that URL isn't really a property of a VCS commit.

If we are going down the "embed url per commit" route, I think we should consider not having this structure at all and instead just having a single `string code_host_url` property on commit and call it a day.

Reply:
I am fine with whatever, was just coding up the proposal :-) we aren't implementing this part now anyways, we can adjust this pre-v1 to whatever you'd like, I can see both arguments (but would probably lean into per-module, now thinking about it)